### ISSUE TITLE
Extremely poor performance (depending on the window size)

### ISSUE THREAD

**Issue by akim-13** [neovide/neovide#1312]
**Bug:**
The bigger the size of the window is (after a certain point), the more unbearable the performance becomes. Full screen is practically unusable.

**To Reproduce:**
1. Open neovide
2. Make it full screen (or make the window >~80% of the screen size)
3. Move the cursor

**Expected behavior:**
Stable performance at 60 fps as it is for smaller window sizes. 

**Demonstration:**
https://user-images.githubusercontent.com/77539913/168424950-0a9273a9-32bd-45ba-8e49-54e2d618206f.mp4

**Desktop:**
 - OS: Arch linux 5.17.5-arch1-1
 - Specs: Intel i7-10750H, NVIDIA GeForce GTX 1650 Ti Mobile, **3840x2400** *(I think this issue might have something to do with 4k resolution)*
 - Neovide Version 0.9.0
 - Neovim Version 0.7.0

**Log:**
[neovide_rCURRENT.log](https://github.com/neovide/neovide/files/8692768/neovide_rCURRENT.log)

**Additional context**
[Neovide (and Neovim) config](https://github.com/akim-13/neovim/blob/master/lua/rc/neovide.lua)

**Comment by akim-13** [comment]
My suspicion about this issue being related to a 4k resolution has been confirmed by [this reply](https://github.com/neovide/neovide/issues/1299#issuecomment-1127603519).

**Comment by vabenil** [comment]
I think I have this same issue. When I run neovide with --multigrid in fullscreen I usually get 60 FPS but when scrolling the FPS can go all the way down to the the 20s, if I reduce the size of the window as you mentioned in your post I get full 60 FPS all the time. When running other neovim guis, such as [nvui](https://github.com/rohit-px2/nvui) I get more than OK performance.

Setup:
- OS: Arch Linux 5.15.54-1-lts
- CPU: Intel(R) Core(TM) i5-5200U CPU @ 2.20GHz
- GPU: Intel Corporation HD Graphics 5500 (integrated graphics)

**Comment by kirillrogovoy** [comment]
I have the same issue on a maxed-out 2019 MacBook Pro 13".

MacOS 12.3.1

**Comment by hatemets** [comment]
Same issue with a 27" QHD screen running Arch Linux. Running the window with the size comprising 1/3 of the screen runs smoothly, anything above that starts to become laggy.

**Comment by ldelossa** [comment]
Is this only present with "--multigrid" ?

**Comment by akim-13** [comment]
> Is this only present with "--multigrid" ?

No

**Comment by kirillrogovoy** [comment]
> Is this only present with "--multigrid" ?

In my case, yes.

**Comment by ldelossa** [comment]
Yes its odd, i have two 4k screens but it seems to only be an issue on my second one if multigrid is used. 

On my main, it doesn't matter, it poor performance either way.

**Comment by fredizzimo** [comment]
I can see som minor stuttering on my 5120x2160 screen, but no where near as bad as demonstrated here, so I can take a stab at fixing it if I have time this weekend.

@ldelossa, is one of your screen using GSync and the other one not?

**Comment by vabenil** [comment]
> Is this only present with "--multigrid" ?

In my case it works fine without `--multigrid`, but performance dies after `--multigrid` and trying to run with full-screen

**Comment by ldelossa** [comment]
@fredizzimo that is correct. One is a Gsync monitor and the other is a basic 4k monitor.

**Comment by fredizzimo** [comment]
@ldelossa Just as a clarification, which of the montitors have good performance, and which has bad?

**Comment by ldelossa** [comment]
The one with gsync

**Comment by fredizzimo** [comment]
Just some quick updates on this. I think I have identified the problems, but I won't promise any time estimates for the fixes. There are rougly 4 different problems

1. The first problem is that not all Windows events are processed before attemptig to render. This can cause events to buffer up, especially on slower systems and everything grinds to a halt.
2. The frame rate limiter does not work very well, and most of the times it uneccesarily limits the frame rates. For example fixing the above problem and disabling all limits causes my frame rates when scrolling to go from around 30 FPS to 130 FPS. On systems where vsync is available, that should be used, but it needs to be fixed without that as well.
3. We use skia `save_layer` to deal with overlapping and transparent windows. That allocates a new `SkSurface` each frame and is very slow. Disabling that, causes the baseline frame rate with not movement to go over 1000 FPS. During scrolling it drops to around 300 FPS, which is starting to look acceptable for my system. The fix would be to pre-allocte these surfaces and re-use them from frame to frame.
4. Some of the effects, including the scrolling is a frame rate dependent. This causes some general choppines, even with high frame rates. Running the effects with a fixed timestep will fix that issue.

Note that my system is fairly powerful with a Ryzen 5 5600X and Gefore RTX 3060, so I'm not really seeing any really bad behaviour. But I believe that most problems descirbed here comes from 1. and 3.

**Comment by fredizzimo** [comment]
I optimized the rendering a bit to not call `save_layer` and the blurring functionality unless there's a transparent floating window that needs it and blurring is not disabled. Standard transparency is fast and causes very little slowdown.

After this the baseline FPS is around 2000, and it drops down to around 1000 (not a typo) when scrolling! A big floating window   with blurring causes the FPS to drop to around 120, and if blurring is disabled to around 1000. So there are some quite big improvments incomming 😄

**Comment by Kethku** [comment]
Amazing. Thats awesome!

**Comment by fredizzimo** [comment]
I just want to say, that I'm still working on this, I didn't have that much time until now, but last weekend I did some more updates. I'm working in this branch, https://github.com/fredizzimo/neovide/commits/fsundvik/improve-performance, which at a given time may work or might not, since I'm quite frequently rebasing and pushing new stuff. Currently it only supports either vsync enable or totally uncapped, and relies on the `--no-idle` flag being set.

At the moment, even if it's completely pixel perfect smooth there's some strange occassional "micro-stuttering". All the deltas the frame dt and the scroll velocity is completely stable when it happens. So, I think it's related to how we deal with the scrolling snapshots, that could cause some gpu stalls, which makes some of the skia updates come too late in the wrong frame. It could also be related to that we don't seem to do proper atomic updates on the rendering state, we should only commit something to the screen when there's a flush.

I have already done some fixes the the scrolling snapshots, but I think I will completely rewrite it now to use a single canvas that is at least twice the size of the Window. Or actually two, one for in progress updates and one that is currently displaying. This simplifies the logic a lot, allows us to add smooth side scroling in the future, and should be much more friendly for the GPU.

**Comment by WillyWonksters** [comment]
Fredizzimo, your experience with this might be relevant to the [Cosmic Text](https://github.com/pop-os/cosmic-text) project that Jeremy Soller and others are working on. Several text handling experts are also discussing the project in the [Iced Discord](https://discord.com/channels/628993209984614400/1027584473631555604) if you are interested.

Anyway, I'm so glad to hear that you are worked on this!

**Comment by tan-wei** [comment]
For me, a 4K monitor with maximize neovide will lag now (0.12.0). Probably I should buy a new computer to solve this. :-P

**Comment by fredizzimo** [comment]
@tan-wei, what is your system specifications?

Neovide should work smoothly on almost all systems. But it's possible that the graphics card drivers are not up to date, or that it selects the wrong GPU or something like that, and causes it to switch to software based rendering, which will be too slow.

**Comment by tan-wei** [comment]
I have several computers, they all shared with a 4K screen:
  1. NUC10 with Windows 10 (so the GPU is Intel HD Graphics, which is not that good)
  2. Mac Mini 2018

I've already tried 0.12.0, neovide is truly much more smooth than the previous version (maybe 0.10? sorry I forget). But still lag ...

As a comparison, I have a quite old Thinkpad T430 (made in 2012 probably), which runs archlinux with a 1920 x 1080 external screen, it is smooth enough. So I think the problem is, my computers  are not that strong.

PS: I also use a virtual machine with default GPU driver, as the README.md says, it is not enough.

**Comment by fredizzimo** [comment]
If you still have problems. would it be possible to provide a tracy log following these instrucitons https://neovide.dev/troubleshooting.html#performance-profiling?

**Comment by dineshKumar777** [comment]
Thanks @crupest  for documentation. It was easy to follow.
@fredizzimo Attaching tracy logs from my system. Please let me know if you need any other info or new logs. Thanks

> System: Mac mini 2018 - 4k monitor - scaled resolution 2560x1440
> recorded events: rapid home-row typing and resizing window

minimal window size logs: 

- In this size, I can use Neovide without slowness but resizing window will cause slowness


Maximised window size logs: 
- Neovide is unusable in my system.

[tracy-logs.zip](https://github.com/neovide/neovide/files/14261495/tracy-logs.zip)
Note: Github not allowing this filetype to upload. so zipped both logs and attached.

**Comment by fredizzimo** [comment]
Can you also provide a neovide log file, that should show what kind of window sizes are drawn.

I suspect that if you have some scaling enabled, then it might draw at the double resolution, and use a lot of memory. In that case it should help to disable the scaling, at least temporarily.

**Comment by fredizzimo** [comment]
I did a bit of digging and the macMini 2018 does indeed have problems with the memory and high resolutions, not just in Neovide, but everywhere https://forums.macrumors.com/threads/mac-mini-2018-bad-ui-performance-with-lg-5k-display.2154518/

That said there might be tricks we could use to decrease the memory consumption.. and at least make it much better.

**Comment by fredizzimo** [comment]
I will do some memory tracking on Windows, where I know the tools to do so, and have it booted up.

The GPU memory usage does seem a bit excessive at 650 MB on a resolution of 3440x1440, so it might be possible to optimize

**Comment by dineshKumar777** [comment]
Thanks for looking into it.

> macMini 2018 does indeed have problems with the memory and high resolutions, not just in Neovide, but everywhere
Sadly Neovide is currently not usable in fullscreen.. My expectation is to have an usable neovide version in fullscreen similar to other editors like vscode or recently updated zed.

attaching gpu_profiled tracy logs with neovide logs.
[tracy-2kscaled-gpuprofile-maximized-window.zip](https://github.com/neovide/neovide/files/14268016/tracy-2kscaled-gpuprofile-maximized-window.zip)


Will attach native 4k resolution logs shortly

**Comment by dineshKumar777** [comment]
wow as soon i enable native 4k resolution i can see the performance improvement.
Especially in minimalized window where neovide typing experience is so smoooooooth. Loved it. 
![image](https://github.com/neovide/neovide/assets/21097695/c1408b44-5c1c-46aa-b16e-9dc3fcc67690)

attaching gpu_profiled tracy logs with neovide logs.

[tracy-4k-gpuprofilied-maximized-window.zip](https://github.com/neovide/neovide/files/14268565/tracy-4k-gpuprofilied-maximized-window.zip)
[tracy-4k-gpuprofile-minimalsized-window.zip](https://github.com/neovide/neovide/files/14268566/tracy-4k-gpuprofile-minimalsized-window.zip)

**Comment by crupest** [comment]
@fredizzimo Would you mind have a look at the tracy log on my machine?

Conditions:
- MacBook Pro with CPU: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
- A 4K screen.
- `cargo run --profile profiling --features gpu_profiling -- --no-fork --no-idle`

[4k-tracy.zip](https://github.com/neovide/neovide/files/14268887/4k-tracy.zip)


Seems like it takes a very very long time to swap buffer.

**Comment by fredizzimo** [comment]
@dineshKumar777, I'm not sure why, but it looks like the gpu profiling did not work. But don't worry, I don't think it would be very useful.

@crupest, yes, it looks almost the same as the logs dineshKumar777 posted, but the problem is much smaller on your machine. Most frames are rendered OK, and then there are a few very slow frames. 

I also checked the memory usage on Windows, and there doesn't seem to be much to optimize regarding the pure memory usage. I can only find around 100 MB allocated, even if Windows shows 650 MB of dedicated GPU memory used. There has to be a reason, because no other application is nearly that big. I suspect it's all the temporary buffers Skia is allocating, but not used for rendering the frame itself. I don't know how to continue the investigation from here, but I will try to figure it out.

**Comment by fredizzimo** [comment]
It definitely seem to be the skia caches, or rather the memory pools for them.

I tried both https://rust-skia.github.io/doc/skia_safe/gpu/type.DirectContext.html#method.set_resource_cache_limit
and 
https://rust-skia.github.io/doc/skia_safe/graphics/fn.set_resource_cache_total_bytes_limit.html

You can try to set them to something smal like `128*1024*1024`. here and see if that improves the situation https://github.com/neovide/neovide/blob/1927e6722e4003153942dc348be97b5d71974e06/src/window/renderer.rs#L73

**Comment by fredizzimo** [comment]
On the other hand, according to tracy, @crupest has 16GB of ram, which makes memory issues much less likely

**Comment by fredizzimo** [comment]
I wonder if we should try to implement the metal API instead of OpenGL? OpenGL has been deprecated on mac for a long time, so that alone might be the reason for the poor performance.

It seems like we could use this to create the context https://github.com/gfx-rs/metal-rs

And Skia  supports Metal directly. 

I will try to pull out the refactoring parts of 
* https://github.com/neovide/neovide/pull/2215

And after that the metal backend could be added the same way as D3D on Windows.

**Comment by fredizzimo** [comment]
One question @crupest, do you run your screen at 60 hz? That's what our vsync system thinks, but if it's out of sync with the real refresh rate then I guess that swap buffers might start blocking (even if opengl vsync is supposed to be turned off in our implementation)

**Comment by fredizzimo** [comment]
One completely unrelated thing that I noticed by chance is that if performance wasn't an issue, then we should enable high resolution buffers https://developer.apple.com/library/archive/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/EnablingOpenGLforHighResolution/EnablingOpenGLforHighResolution.html

That should make the fonts sharper on especially retina displays.

**Comment by fredizzimo** [comment]
Now, when I look at this https://developer.apple.com/library/archive/documentation/GraphicsImaging/Conceptual/OpenGL-MacProgGuide/opengl_designstrategies/opengl_designstrategies.html#//apple_ref/doc/uid/TP40001987-CH2-SW12

I think our displaylink callback is done slightly wrong. First of all, they enable vsync, while we turn it off. Secondly it seems like we should not return from the callback until the frame is rendered.

So, I think we need a two-way signal, `request_redraw`, which is already there in the d3d branch, and a redraw done signal. If there's no redraw request the frame calback can return immediately, otherwise it should wait until we have rendered the screen.

So, these fixes are probably the first ones that we should do, before trying Metal.

**Comment by fredizzimo** [comment]
I have now extracted the refactoring parts of the d3d branch into its own PR
* https://github.com/neovide/neovide/pull/2359

So, any experiments with either the vsync/display link or the rendering backend should be built on top of that.

**Comment by crupest** [comment]
@fredizzimo These are great work of you.

First, let me answer your questions,

> On the other hand, according to tracy, @crupest has 16GB of ram, which makes memory issues much less likely

Yes, I have 16GB ram.

> One question @crupest, do you run your screen at 60 hz? That's what our vsync system thinks, but if it's out of sync with the real refresh rate then I guess that swap buffers might start blocking (even if opengl vsync is supposed to be turned off in our implementation)

Yes, 60Hz.

Second, please allow me to do some conclusion for macos,

1. Limit skia cache size. I will try this to get a result very soon.
2. Use `setWantsBestResolutionOpenGLSurface` API to enable high resolution buffers for OpenGL. I don't how easy this is, because it revolves coordinate conversion for dpi. I'm not sure if our current implementation is already able to handle this. Will try this second. Otherwise, we have to do some work on this if we REALLY want to use it.
3. Correct the implementation of vsync and DisplayLink. Looks like this is much harder than above. But compared to Metal, this is more gentle and less destructive, maybe?
4. The ultimate goal: Metal, like D3D. Actually I knew Apple has abandoned OpenGL for a long time.

**Comment by fredizzimo** [comment]
@crupest, I think you can skip 2. for now. It should be a quality improvement, not a performance one. And it actually defaults to true https://developer.apple.com/documentation/appkit/nsview/1414938-wantsbestresolutionopenglsurface

But setting it to false might help in @dineshKumar777's case with scaling enabled, but with more blurry fonts as a compromise.

**Comment by crupest** [comment]
@fredizzimo The test result of limiting resource cache size outs.

- `4k-limit-128-tracy` 4k and limit to 128MB
- `4k-limit-64-tracy` 4k and limit to 64MB, same for 32
- `4k-tracy` 4k and no limit
- `limit-128-tracy` not 4k and limit to 128MB

[4k-limit-tracy.zip](https://github.com/neovide/neovide/files/14277450/4k-limit-tracy.zip)

As I can see, there is only little improvement. Swapping buffer seems to be a little faster.

The real effect is it is still very laggy. (The frame rate can't keep 60.) And 64 and 32 compared to 128 has NO improvement. Looks like as my computer turns on for some time and it slows down due to heat, the performance is not better than no limit.

**Comment by ctoLuu** [comment]
Has this issue been resolved?I have a poor experience on a 4K 3640 * 2140 monitor, which is very laggy on a large screen but not a problem on a small screen. if i set --no-multigrid, the situation will improve but it will still be very laggy.

**Comment by crupest** [comment]
@ctoLuu Hi!

You could try this pr https://github.com/neovide/neovide/pull/2461 if you are willing. It implements Metal renderer on macOS. Since OpenGL has been deprecated by Apple and Metal supersedes it, the performance of Metal might be better than OpenGL.

The pr is not completed, because I'm totally new to macOS and Metal development, and the implementation is very immature. However I have been using it personally for a long time and never run into problem. If you use it and find a bug, I'll be very willing to hear the report from you.

One thing to notice. I work on a Mac Book Pro with *Intel* CPU. Though the performance is actually a little better with Metal, it is still much NOT enough to run Neovide fluently in fullscreen on 4K screen. So I always use a smaller window. I don't know the performance of your hardware, but I have to remind you of this situation.

At last, you can pull the pr and build it by yourself. If you know about Rust build system, it should be relatively simple. The pr is already synced with main branch. You can also download a built one directly on CI at https://github.com/neovide/neovide/actions/runs/11305856557?pr=2461#artifacts .

Hope this can help you!

**Comment by crupest** [comment]
Sorry that I forget this  bug is on Linux. Just ignore this. 😅

**Comment by ctoLuu** [comment]
@crupest Yes, I found this problem on ubuntu22.04, but thanks for your answer. :smiley:

**Comment by bassamsdata** [comment]
> ou can also download a built one directly on CI at https://github.com/neovide/neovide/actions/runs/11305856557?pr=2461#artifacts .

@crupest Thank you for the PR and the detailed comment! I had an annoying issue with Neovide lagging heavily when going full screen on my MacBook Pro. After installing the build you mentioned, it now performs much better in full screen mode on my MacBook.

It’s not perfect yet, but I’d say there’s at least a 50% performance improvement, making it much more usable. Thanks again!

**Comment by crupest** [comment]
@bassamsdata Glad to hear that! If you run into some bug, feel free to leave a comment in that pr. I'll appreciate it and try to fix the bugs!

**Comment by jmaliksi** [comment]
@crupest I pulled your PR and built it on my 2019 intel 13" macbook hooked up to a 4k monitor and agree with your assessment that while it's a bit better, it's still not viable on fullscreen. The PR looks very promising regardless, so thanks for all the effort!

@fredizzimo as an additional data point, the full screen lag only happens if I have the display scaling in os settings set to anything ''under'' 3840x2160 (native resolution)--aside from the scaling to 1920x1080 (the smallest) and my guess is that's just because 1080p is easy to render. For other resolutions, my hunch is that the OS, or skia, or something (I'm very unfamiliar with application rendering, sorry for nebulous language) is causing the lag as it re-scales each frame.

So the workaround of basically disabling scaling in OS settings has worked for me so far, just annoying that I have to rescale all my other apps

**Comment by daniel-odrinski** [comment]
One interesting observation from my testing today - the OpenGL implementation is way smoother than the DirectX one on Windows. Whilst both are stuttering, the DirectX implementation is noticeably worse. You would think that the opposite would be true. For example, the Metal renderer is noticeably smoother than the OpenGL one on macOS, almost to the point of being perfect.

4096x2160 @ 120Hz
RTX 3070 Laptop GPU
Windows 11
Neovide 0.14.0

**Comment by fredizzimo** [comment]
For me and many others the D3D implementation is much smoother. are you sure your drivers are up to date?

It's also possible that Neovide selects the wrong and different GPU with D3D. At least that seems to be the case for me, it selects the Radeon 780M GPU, rather than my RTX 4070. But I haven't investigated deeper, since I just got this laptop and it's fast enough to render almost without frame drops at 165 Hz, 2560x1440.

You can check which GPU it actually uses in the task manager

**Comment by daniel-odrinski** [comment]
Hi @fredizzimo , thanks for the comment.

Yep, I went into Windows 11's Graphics settings and specifically set Neovide to use the RTX 3070 instead of the integrated AMD Radeon graphics. I also confirmed in Task Manager that it was using it the NVIDIA GPU. Otherwise, it was selecting the AMD Graphics by default.

I will make sure that my drivers are updated to the latest version.

**Comment by fredizzimo** [comment]
@daniel-odrinski. If you are able to, you could try capturing a performance profiling trace https://neovide.dev/troubleshooting.html#performance-profiling

That should give us much more information about what's wrong.

**Comment by daniel-odrinski** [comment]
@fredizzimo Awesome, will do - thank you 👍🏻
